source/_posts/2018-01-31-infrastrucutre-as-code-a-beginners-perspective.markdown

```diff
@@ -0,0 +1,123 @@
+
+<!--more-->
+
+## What does Infrastructure as Code mean?
+
+As the name might suggest you, IAC is basically writing code that describes your infrastructure, just as if you were writing a program in a high level language. This allows you to automate de process of provisioning infrastructure using tools proven to work.
+
+## How does it help you?
+
+This allows you to automate, test and reuse the code for the infrastructure you are working on. You can also have versioning of your infrastructure code and check it in version control alongside with the code it will be serving. You can do code reviews on it as well, testing, and all the good practices you follow with your codebase.
+
+**You might be thinking:** *Ok, that seems like good things to have, but can you please give me a concrete use case?*
+
+I will give you two 😁
+
+### Case 1 - Simple web app_1, app_2, app_3, ..., app_n
+
+Imagine you are working for a company that builds basic web apps for their clients, so, what does a basic web app needs?
+
+Probably a web server and a database (I am oversimplifying), so every time you have a new client for a web app you go to AWS (or whatever cloud provider you use) and with some clicking around you get 2 VMs, one for running your app, and another one for the database.
+
+Oh wait, I forgot, you need two copies (at least) of this infrastructure, one for production and one for staging.
+
+Now, you have two copies of the same infrastructure. You can start deploying your code for that client.
+
+Two months go by and you have another client that needs the same infrastructure, and there you go again, you click around in AWS and get that working.
+
+With IAC you can have a module that does all the heavy lifting for you, you run that code and in a couple of minutes you are good to go. You need to update all the infrastructure of your clients? No problem just update the module, run it and you are good to go!
+
+If you have been doing this by hand in AWS, then you are probably already seeing the benefits of having IAC. But if you don't believe me yet, just read the next case.
+
+### Case 2 - Going crazy with a complex app
+
+You came up with a great idea, a really cool web app, it started off simple, but now you have a lot of clients that rely on that app. In this case one web server and one database are not going to be enough.
+
+You start by adding one load balancer and another web server, you write the process down in a document. Some time goes by and you need to add another server, you follow the process in the document and you are done.
+
+Next month you realize that only one database server it not enough and you add a replica.
+
+Next you have to move some workload out of the main servers, you need a message queue and a server to run workers, you do it by hand, but at this point you forget to update the documents.
```

Comment:
What do you think about using a full stop instead of a comma after *and a server to run workers*?